exporter/trace/stackdriver/stackdriver.go

```diff
@@ -162,6 +186,71 @@ func (e *Exporter) uploadToStackdriver(spans []*trace.SpanData) {
 	}
 }
 
+// FromRequest extracts a Stakdriver Trace span context from incoming requests.
+func (e *Exporter) FromRequest(req *http.Request) (sc trace.SpanContext, ok bool) {
+	h := req.Header.Get(httpHeader)
+	// See https://cloud.google.com/trace/docs/faq for the header format.
+	// Return if the header is empty or missing, or if the header is unreasonably
+	// large, to avoid making unnecessary copies of a large string.
+	if h == "" || len(h) > 200 {
+		return trace.SpanContext{}, false
+	}
+
+	// Parse the trace id field.
+	slash := strings.Index(h, `/`)
+	if slash == -1 {
+		return trace.SpanContext{}, false
+
+	}
+	tid, h := h[:slash], h[slash+1:]
+
+	buf, err := hex.DecodeString(tid)
+	if err != nil {
+		return trace.SpanContext{}, false
+	}
+	copy(sc.TraceID[:], buf)
+
+	// Parse the span id field.
+	spanstr := h
+	semicolon := strings.Index(h, `;`)
+	if semicolon != -1 {
+		spanstr, h = h[:semicolon], h[semicolon+1:]
+	}
+	sid, err := strconv.ParseUint(spanstr, 10, 64)
+	if err != nil {
+		return trace.SpanContext{}, false
+	}
+
+	buf = make([]byte, 8)
+	binary.PutUvarint(buf, sid)
+	copy(sc.SpanID[:], buf)
+
+	// Parse the options field, options field is optional.
+	if !strings.HasPrefix(h, "o=") {
+		return sc, true
+
+	}
+	o, err := strconv.ParseUint(h[2:], 10, 64)
+	if err != nil {
+		return trace.SpanContext{}, false
+
+	}
+	sc.TraceOptions = trace.TraceOptions(o)
+	return sc, true
+}
+
+// ToRequest modifies the given request to include a Stackdriver Trace header.
+func (e *Exporter) ToRequest(sc trace.SpanContext, req *http.Request) {
+	var b bytes.Buffer
+	b.WriteString(hex.EncodeToString(sc.TraceID[:]))
+	sid, _ := binary.Uvarint(sc.SpanID[:])
+	b.WriteString("/")
+	b.WriteString(strconv.FormatUint(sid, 10))
+	b.WriteString(";o=")
+	b.WriteString(strconv.FormatInt(int64(sc.TraceOptions), 10))
```

Comment:
Would it perhaps be more clear to avoid all this calls but do them once as
```go
sid, _ := binary.Uvarint(sc.SpanID[:])
fmt.Fprintf(buf, "%x/%d;o=%d", sc.TraceID[:], sid, int64(sc.TraceOptions))
```

Reply:
Done